listing/listing.go

```diff
@@ -0,0 +1,195 @@
+	// items from the underlying source and the request fails, even if there
+	// are no more items to iterate over. In this case, Next will return the
+	// error.
+	HasNext(context.Context) bool
+
+	// Next returns the next item in the iterator. If there are no more items
+	// to iterate over, it returns ErrNoMoreItems. If there was an error
+	// fetching the next page of items, it returns that error. Once Next
+	// returns ErrNoMoreItems or an error, it will continue to return that
+	// error.
+	Next(context.Context) (T, error)
+}
+
+// ToSlice returns all items from the iterator as a slice. If there was an
+// error fetching items at any time, it returns that error.
+func ToSlice[T any](ctx context.Context, it Iterator[T]) ([]T, error) {
+	var items []T
+	for it.HasNext(ctx) {
+		item, err := it.Next(ctx)
+		if err != nil {
+			return nil, err
+		}
+		items = append(items, item)
+	}
+	return items, nil
+}
+
+// Use struct{} for Req for iterators that don't need any request structure.
+type iteratorImpl[Req, Resp, T any] struct {
+	// nextReq is the request to be used to fetch the next page. If nil, then
+	// there is no next page to fetch.
+	nextReq *Req
+
+	// getNextPage fetches the next page of items, returning the deserialized
+	// response and error.
+	getNextPage func(context.Context, Req) (Resp, error)
+
+	// getItems selects the items being iterated over from the response.
+	getItems func(Resp) []T
+
+	// getNextReq is used to get the next request to be used in the next page.
+	// If nil, then there is no next page to fetch.
+	getNextReq func(Resp) *Req
+
+	// currentPage is the current page of items.
+	currentPage []T
+
+	// currentPageIdx is the index of the next item from currentPage to return.
+	currentPageIdx int
+
+	// lastErr is the last error returned by getNextPage.
+	lastErr error
+}
+
+var ErrNoMoreItems = errors.New("no more items")
+
+// Returns a new iterator. The iterator will fetch the next page of items
+// lazily, when needed. nextReq is the request to be used to fetch the initial
+// page. If nil, then no page will be fetched. getNextPage fetches the next
+// page of items, returning the deserialized response and error. getItems
+// selects the items being iterated over from the response. getNextReq is used
+// to get the next request to be used in the next page. If the returned value
+// is nil, then there is no next page to fetch.
+func NewIterator[Req, Resp, T any](
+	nextReq *Req,
+	getNextPage func(context.Context, Req) (Resp, error),
+	getItems func(Resp) []T,
+	getNextReq func(Resp) *Req,
+) *iteratorImpl[Req, Resp, T] {
+	return &iteratorImpl[Req, Resp, T]{
+		nextReq:     nextReq,
+		getNextPage: getNextPage,
+		getItems:    getItems,
+		getNextReq:  getNextReq,
+	}
+}
+
+func (i *iteratorImpl[Req, Resp, T]) loadNextPageIfNeeded(ctx context.Context) error {
+	if i.currentPageIdx < len(i.currentPage) {
+		return nil
+	}
+	if i.nextReq == nil {
+		i.currentPage = nil
+		i.currentPageIdx = 0
+		return nil
+	}
```

Comment:
Since we only set `nextReq` from this function, we can reset + return when we assign it.

Also, if you don't move this one, we'd need to move the check for `i.lastErr` up here. It should take precedence over the `lastErr` check.